### DISCUSSION TITLE
Telnet not working : on VSCPD

### DISCUSSION THREAD

**Discussion by dinguluer** [grodansparadis/vscp#271] [Q&A] [answered]
Hi ,

How to use telnet on vscpd.

killed the vscpd.. then it restarted again. :--
![image](https://github.com/grodansparadis/vscp/assets/370819/8e2e0738-7045-4082-93cf-54ed69ee3ec9)

then i tried to connect using telnet 
![image](https://github.com/grodansparadis/vscp/assets/370819/a71bad51-d488-4a04-a73b-0ddd0f5667b3)

Getting this output, from vscpd status :--
![image](https://github.com/grodansparadis/vscp/assets/370819/9fef5242-f823-453e-bca6-300de52eeee4)

Driver is here :--
![image](https://github.com/grodansparadis/vscp/assets/370819/37739e8b-8f51-47da-96ac-c03270f65ceb)
![image](https://github.com/grodansparadis/vscp/assets/370819/c2054eb1-5784-416f-9fe8-b550d33dc5ac)

Vscpd.json file :--
      "path-driver": "/var/lib/vscp/drivers/level2/vscpl2drv-tcpiplink.so",
      "path-config": "/usr/share/vscpl2drv-tcpiplink/tcpiplink.json",
![image](https://github.com/grodansparadis/vscp/assets/370819/0c39b357-6c87-4e22-a9f6-62eb5098a549)

Permission is denied to level2 folder :--
i did   --> chmod 666 level1 level2
![image](https://github.com/grodansparadis/vscp/assets/370819/8b1e5902-576e-48be-be8f-64de762184f2)

May i know what is missing from my end now ?

Regards,
Dinesh

**Comment by dinguluer** [comment]
Hi AKe ,

You said :--
1>the link driver is for connecting to another VSCP tcp/ip link host. 
2> The srv driver mentioned earlier is the driver you are after if you want a telnet interface.
3> Easiest is to use the cansocket driver and create a virtual interface. Then you can use the can-tools

What does it means ?

Regards,
Dinesh

**Reply by grodansparadis** [reply]
You said :--
1>the link driver is for connecting to another VSCP tcp/ip link host.
2> The srv driver mentioned earlier is the driver you are after if you want a telnet interface.

What does it means ?

**Comment by grodansparadis** [comment, marked as answer]
There are two drivers that use VSCP tcp/ip link protocol

[vscpl2drv-tcpiplink](https://github.com/grodansparadis/vscpl2drv-tcpiplink)

vscpd <-> vscpd or oyther device with VSCP tcp/ip link interface 

and

[vscpl2drv-tcpipsrv](https://github.com/grodansparadis/vscpl2drv-tcpipsrv)

VSCP tcp/ip link server.    vscpd <-> clients

**Reply by dinguluer** [reply]
So you mean to say.

Use this driver for vscpd to vscpd communication right ? :--
[vscpl2drv-tcpiplink](https://github.com/grodansparadis/vscpl2drv-tcpiplink)

Use this driver for client to vscpd communication right ? :--
[vscpl2drv-tcpipsrv](https://github.com/grodansparadis/vscpl2drv-tcpipsrv)

**Reply by grodansparadis** [reply]
yes
vscpl2drv-tcpiplink - vscpd to connect to another unit that talk VSCP tcp/ip protocol.
vscpl2drv-tcpipsrv  - Server for vscpd that let clients connect using VSCP tcp/ip protocol

**Reply by dinguluer** [reply]
So I need to use this for telnet, right :--
vscpl2drv-tcpipsrv - Server for vscpd that let clients connect using VSCP tcp/ip protocol

Also :--
What is purpose of this MQTT node to be added inside the vscpd.json file with tcpip-srv ?
And what is purpose of two port 1883 & 9958 ?

![image](https://github.com/grodansparadis/vscp/assets/370819/7d6324c2-8360-4a24-87c6-7a66e7646622)

![image](https://github.com/grodansparadis/vscp/assets/370819/ea1aba39-dd03-4882-99f8-07530a7895ca)

**Reply by grodansparadis** [reply]
If you want to connect manually yes telnet is right.   
If you want to connect using a program use [vscp-helper-lib ](https://github.com/grodansparadis/vscp-helper-lib)

The VSCP driver connects to a MQTT server.  This is how all drivers connect nowadays.  In the past all drivers was talking to each other inside the vscpd server. Now this can be accomplish by using the same topic on an MQTT server.

**Comment by dinguluer** [comment]
Also which driver VSCPD uses to communicate to the MQTT broker here ? :--


![image](https://github.com/grodansparadis/vscp/assets/370819/178ffa64-8237-4a53-b75a-0cfde6397e80)

**Reply by grodansparadis** [reply]
All drivers talk to a MQTT broker.    driver <-> vscpd <-> MQTT       
vscpd now just interface the driver and provide the communication with MQTT brokers.

**Comment by dinguluer** [comment]
What is communication medium, between two vscpl2drv-tcpiplink  --> is it TCP/IP  ? :--

![image](https://github.com/grodansparadis/vscp/assets/370819/f354b0df-a72d-43cc-b906-44424b3db6f7)

**Reply by grodansparadis** [reply]
yes tcp/ip

**Comment by dinguluer** [comment]
What is this port used by every driver, in MQTT node in vscpd.json file :--
"port": 1883,

**Reply by grodansparadis** [reply]
VSCP tcp/ip link has 9598 registered with ICANN as it's standard port.  https://en.wikipedia.org/wiki/Registered_port

**Reply by dinguluer** [reply]
Same driver --> vscpl2drv-tcpipsrv  --> what is purpose of two port 1883 & 9958 ?

1883 port is used by whom ?

**Reply by grodansparadis** [reply]
Sorry i meant 9598 is used by VSCP tcp/ip link protocol. [](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt)

1883 is MQTT. The mosquitto test server has a good list of MQTT standard ports used. [](https://test.mosquitto.org/)  As you see 12883 is usually used for unencrypted traffic.

**Reply by dinguluer** [reply]
ok got

**Comment by dinguluer** [comment]
Also what exactly is this log, TCPIPlink not available ?
**systemctrl status vscpd**  :--

![image](https://github.com/grodansparadis/vscp/assets/370819/3392231e-be26-452f-b64e-d2e51c8ee2e9)

**Reply by grodansparadis** [reply]
Is  the driver available in the folder?

**Reply by dinguluer** [reply]
Yes its here :--
![image](https://github.com/grodansparadis/vscp/assets/370819/d5ba8b42-298d-453e-8989-948b7562b2d4)


What is the right process to stop/kill vscpd & restart it, so that it takes latest settings in vscpd.json file ?

**Reply by grodansparadis** [reply]
Then it can be a privilege problem.  [See here](https://github.com/grodansparadis/vscp/issues/269)

**Reply by dinguluer** [reply]
Means no access permission ?

**Reply by grodansparadis** [reply]
Yes. It should be as in the image in the link.  If you don't care for security just make the folders and files available for everyone under /var/lib/vscp

**Reply by dinguluer** [reply]
sudo chmod -R 777 /var/lib/vscp

this command ?


or I shall make same permission as shown in this image :--
![image](https://github.com/grodansparadis/vscp/assets/370819/1c96ba3b-5b8d-407a-b1d6-5e5f4fe7a0b4)

**Reply by grodansparadis** [reply]
yes that works

**Reply by dinguluer** [reply]
I run the chmod commsnd :--
![image](https://github.com/grodansparadis/vscp/assets/370819/d9a433cd-cd46-4d09-9c68-c2d4ef85c3f8)

I killed vscpd & it again started :--
![image](https://github.com/grodansparadis/vscp/assets/370819/806fd6f9-37d4-4acf-a9d6-02371ab08671)

**sudo systemctl status vscpd :--**
![image](https://github.com/grodansparadis/vscp/assets/370819/345b7dfb-0542-4ae7-8e35-715cf9ca4495)

Please suggest have i missed something in process ?

**Reply by grodansparadis** [reply]
Its says the path does not exists so it must be something there.  What does 

ls -la /var/lib/vscp/drivers/level2

give?

**Reply by dinguluer** [reply]
>> ls -la /var/lib/vscp/drivers/level2 :---

![image](https://github.com/grodansparadis/vscp/assets/370819/10ea3cf5-3429-4b01-96b0-519cd03bc183)

**Reply by grodansparadis** [reply]
Note the "lib" prefix in the name

**Reply by dinguluer** [reply]
What shall i do in this case ?
**Shall i rename the file ?**

built from source -- vscpl2drv-tcpiplink :--
I installed this :--
sudo apt-get install libcurl4-openssl-dev

Then compile make & make install :--
Install the project...
-- Install configuration: ""
**-- Installing: /var/lib/vscp/drivers/level2/libvscpl2drv-tcpiplink.so**
-- Installing: /usr/share/vscpl2drv-tcpiplink/tcpiplink.json
-- Installing: /usr/share/man/man1/vscpl2drv-tcpiplink.1

![image](https://github.com/grodansparadis/vscp/assets/370819/ff3ff7c0-50e6-4047-afd4-48dec5d56e61)

**Reply by dinguluer** [reply]
I renamed the file it worked :---
![image](https://github.com/grodansparadis/vscp/assets/370819/fdf0d68a-9ba0-483c-9c67-3ec39737796f)

**Reply by grodansparadis** [reply]
just add "lib" in the config file before the filename.

**Reply by grodansparadis** [reply]
Yes that work to. It is wrongly named now so that is maybe the best way.

**Reply by dinguluer** [reply]
Thanks

**Comment by dinguluer** [comment]
You told :--
If you want to connect using a program use [vscp-helper-lib](https://github.com/grodansparadis/vscp-helper-lib)

**here program means C program ?**

As per this link :--
https://grodansparadis.github.io/vscp-helper-lib/#/README
The library is currently available for [c/c++](https://www.vscp.org/#download), [python](https://github.com/grodansparadis/pyvscp) and [node.js](https://grodansparadis.github.io/vscp-helper-lib/#/) and is available on Unix/Linux and Windows. It should be easily ported to Mac and other systems.

**Reply by grodansparadis** [reply]
yes, c/c++ or python and node.js.  It is just a c-lib so easy to port to other languages.  It handles the VSCP tcp/ip link protocol and contains a lot of other helpers.  [Docs are here](https://docs.vscp.org/#vscphelper)

**Comment by dinguluer** [comment]
https://mqtt.org/faq/
Yes. TCP/IP port 1883 is reserved with IANA for use with MQTT. 
TCP/IP port 8883 is also registered, for using MQTT over SSL.

![image](https://github.com/grodansparadis/vscp/assets/370819/a3269f00-30cd-4f64-8750-745c315062aa)

Got it

**Comment by dinguluer** [comment]
Why by default vscpd do not have entry for **tcpip-srv**  in /etc/vscp/ vscpd.json file ?

So now how to proceed with TELNET connection ?

Compile this from source and install :--
https://github.com/grodansparadis/vscpl2drv-tcpipsrv

And Create entry in :--
/etc/vscp/ vscpd.json

right ?

**Reply by grodansparadis** [reply]
Why not just install driver?

**Reply by dinguluer** [reply]
ok will try :--
No link to download deb :--
https://github.com/grodansparadis/vscpl2drv-tcpipsrv#install-the-driver-on-linux

nxf77952@lsv03657:~$ uname -a
Linux 5.4.0-153-generic #170-Ubuntu SMP Fri Jun 16 13:43:31 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux


will it work on my machine ?

**Reply by grodansparadis** [reply]
https://github.com/grodansparadis/vscpl2drv-tcpipsrv/releases/tag/v15.0.0

**Comment by dinguluer** [comment]
Also where to get the details of , these parameters mentioned inside the **MQTT** node ?:---
![image](https://github.com/grodansparadis/vscp/assets/370819/9971eb95-f22f-41af-b599-b581a827649f)

**Reply by grodansparadis** [reply]
In the vscpd manual

**Reply by grodansparadis** [reply]
https://docs.vscp.org/#vscpd

**Reply by dinguluer** [reply]
This link do not have all the details :--
https://docs.vscp.org/vscpd/13.1/#/level2_driver_mqtt


Please can you share right link ?

**Reply by dinguluer** [reply]
Ok many documentation this is best :--
https://grodansparadis.github.io/vscp/#/configuring_the_vscp_daemon
https://grodansparadis.github.io/vscp/#/configuring_the_vscp_daemon?id=config-mqtt
https://grodansparadis.github.io/vscp/#/publishing_server

**Comment by grodansparadis** [comment]
Thats for the MQTT driver. No need for it today.

Settings are here  https://github.com/grodansparadis/vscpl2drv-tcpipsrv and is the same as for the daemon config file https://grodansparadis.github.io/vscp/#/configuring_the_vscp_daemon?id=config-mqtt

**Comment by dinguluer** [comment]
Why every driver has its own GUID :--
![image](https://github.com/grodansparadis/vscp/assets/370819/90da767e-0976-432d-aae2-ff419d82620b)

Ideally GUID should be of HARDWARE node devices.

**Reply by grodansparadis** [reply]
It is used to identify the driver
